peer-ids/peer-ids.md

```diff
@@ -185,6 +185,19 @@ To sign a message, we hash the message with SHA 256, and then sign it with the
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
```

Comment:
This key is failing in nim-libp2p, here:
https://github.com/status-im/nim-libp2p/blob/a1eb53b1813df5d4fa1b343ef003ad9fb5ebb1d8/libp2p/crypto/ecnist.nim#L619-L620

I'm not familiar enough to know if it's a problem on our side or an invalid key, but can look deeper if no one knows

Reply:
Looked a bit more into it
https://www.rfc-editor.org/rfc/rfc5915
Specificies:
```asn1
   ECPrivateKey ::= SEQUENCE {
     version        INTEGER { ecPrivkeyVer1(1) } (ecPrivkeyVer1),
     privateKey     OCTET STRING,
     parameters [0] ECParameters {{ NamedCurve }} OPTIONAL,
     publicKey  [1] BIT STRING OPTIONAL
   }
```
>    version specifies the syntax version number of the elliptic curve
      private key structure.  For this version of the document, it SHALL
      be set to ecPrivkeyVer1, which is of type INTEGER and whose value
      is one (1).

But this key contains `0`:
https://lapo.it/asn1js/#MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgEi8z2tSed90aRnrEtLlDIHnTrAwRD5pgjob_MadVLlWhRANCAAStpBjE8fdxXvGjZdmuHoD11waatV7hnxxNd7ROqNycvaVY_ojcMnecaFc78zDa7q4fRBWKO31LMlIwqNrw1G3F
(uncheck "with definitions", first integer = 0)

So unless another spec is used, this is invalid

Reply:
The first integer being `0` is defined here in [RFC5208](https://www.rfc-editor.org/rfc/rfc5208)
```
PrivateKeyInfo ::= SEQUENCE {
  version                   Version,
  privateKeyAlgorithm       PrivateKeyAlgorithmIdentifier,
  privateKey                PrivateKey,
  attributes           [0]  IMPLICIT Attributes OPTIONAL 
}
```
> version is the syntax version number, for compatibility with
      future revisions of this document.  It shall be 0 for this version
      of the document.

as a part of PKCS#8 "header".  
The `1` you are looking for is the next integer
```
SEQUENCE (3 elem)
  INTEGER 0 ---- Version
  SEQUENCE (2 elem) ---- PrivateKeyAlgorithmIdentifier 
    OBJECT IDENTIFIER 1.2.840.10045.2.1 ecPublicKey (ANSI X9.62 public key type)
    OBJECT IDENTIFIER 1.2.840.10045.3.1.7 prime256v1 (ANSI X9.62 named elliptic curve)
  OCTET STRING (109 byte) 306B0201010420122F33DAD… ---- PrivateKey
    SEQUENCE (3 elem)
      INTEGER 1    <<<< here
      OCTET STRING (32 byte) 122F33DAD49E77DD1A... ---- 32 bytes private key
      [1] (1 elem)
        BIT STRING (520 bit) 000001001010110110100... ---- 520 bit or 65 bytes public key
```
which is `1`.  
Additionally, the key generated by Rust implementation of `secp256r1` also generates keys with first integer being `0`.

Reply:
The problem is I don't know what "DER-encoded PKIX" actually means. The key itself is valid, but I'm not sure whether the wrapper is correct.

Reply:
Now I believe the spec refers to RFC5915. I don't know about the encoding, but I guess `306B0201010420122F33DAD49E77DD1A467AC4B4B9432079D3AC0C110F9A608E86FF31A7552E55A14403420004ADA418C4F1F7715EF1A365D9AE1E80F5D7069AB55EE19F1C4D77B44EA8DC9CBDA558FE88DC32779C68573BF330DAEEAE1F44158A3B7D4B325230A8DAF0D46DC5` works? However the `privateKey` field now also contains public key.

Reply:
> > I think go-libp2p does too, since we use some test vectors generated from their lib, and they look like this:
> 
> Can you update this PR with those test vectors? I had no idea we already had some!

Or point to them where they are?

Reply:
We just have some that we generated from go-libp2p here: https://github.com/status-im/nim-libp2p/blob/a1eb53b1813df5d4fa1b343ef003ad9fb5ebb1d8/tests/testpeerid.nim

Reply:
> We just have some that we generated from go-libp2p here: https://github.com/status-im/nim-libp2p/blob/a1eb53b1813df5d4fa1b343ef003ad9fb5ebb1d8/tests/testpeerid.nim 

Awesome thanks!

I'll use those for this PR then!

@drHuangMHT Can you use those for our tests?

Reply:
Failed to decode provided keys in https://github.com/status-im/nim-libp2p/blob/a1eb53b1813df5d4fa1b343ef003ad9fb5ebb1d8/tests/testpeerid.nim. However by skipping the first 4 bytes the keys can be decoded, which is the key in
> You are right, nim-libp2p expects the PrivateKey directly. I think go-libp2p does too, since we use some test vectors generated from their lib, and they look like this: https://lapo.it/asn1js/#MHcCAQEEID5bH-lxLmwxSUKnUL1nSF3jwe_oWxv7Ugro-a49-kpMoAoGCCqGSM49AwEHoUQDQgAE3j0wD6Nq4Oj11TCJnYOrq0Sr8xYfFipLyQHY5uzaAg6LbV-NowUl5x1oUVEMCY5cR8ZGpZf7Tc7ANOn3fECeYg  

The online decoder also can't decode those keys.

Reply:
Never mind the first 4 bytes come from protobuf encoding